File: testers/avro/scala/generated-and-checked-in/com/example/events/producer/CustomerOrderProducer.scala
package com.example.events.producer

import com.example.events.CustomerOrder
import com.example.events.header.StandardHeaders
import java.lang.AutoCloseable
import java.util.concurrent.Future
import org.apache.kafka.clients.producer.Producer
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.clients.producer.RecordMetadata

/** Type-safe producer for customer-order topic */
case class CustomerOrderProducer(
  producer: Producer[String, CustomerOrder],
  topic: String = "customer-order"
) extends AutoCloseable {
  /** Send a message to the topic */
  def send(
    key: String,
    value: CustomerOrder
  ): Future[RecordMetadata] = {
    return producer.send(new ProducerRecord[String, CustomerOrder](topic, key, value))
  }

  /** Send a message with headers to the topic */
  def send(
    key: String,
    value: CustomerOrder,
    headers: StandardHeaders
  ): Future[RecordMetadata] = {
    return producer.send(new ProducerRecord[String, CustomerOrder](
      topic,
      null,
      key,
      value,
      headers.toHeaders
    ))
  }

  /** Close the producer */
  override def close: Unit = {
    producer.close
  }
}

File: testers/avro/java-async/generated-and-checked-in/com/example/events/producer/CustomerOrderProducer.java
package com.example.events.producer;

import com.example.events.CustomerOrder;
import com.example.events.header.StandardHeaders;
import java.util.concurrent.CompletableFuture;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;

/** Type-safe producer for customer-order topic */
public record CustomerOrderProducer(Producer<String, CustomerOrder> producer, String topic)
    implements AutoCloseable {
  public CustomerOrderProducer(Producer<String, CustomerOrder> producer) {
    this(producer, "customer-order");
  }

  public CustomerOrderProducer withProducer(Producer<String, CustomerOrder> producer) {
    return new CustomerOrderProducer(producer, topic);
  }

  public CustomerOrderProducer withTopic(String topic) {
    return new CustomerOrderProducer(producer, topic);
  }

  /** Close the producer */
  @Override
  public void close() {
    producer.close();
  }

  /** Send a message to the topic asynchronously */
  public CompletableFuture<RecordMetadata> send(String key, CustomerOrder value) {
    return CompletableFuture.supplyAsync(
            () -> {
              CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
              producer.send(
                  new ProducerRecord<String, CustomerOrder>(topic, key, value),
                  (result, exception) -> {
                    if (exception != null) {
                      future.completeExceptionally(exception);
                    } else {
                      future.complete(result);
                    }
                  });
              return future;
            })
        .thenCompose(f -> f);
  }

  /** Send a message with headers to the topic asynchronously */
  public CompletableFuture<RecordMetadata> send(
      String key, CustomerOrder value, StandardHeaders headers) {
    return CompletableFuture.supplyAsync(
            () -> {
              CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
              producer.send(
                  new ProducerRecord<String, CustomerOrder>(
                      topic, null, key, value, headers.toHeaders()),
                  (result, exception) -> {
                    if (exception != null) {
                      future.completeExceptionally(exception);
                    } else {
                      future.complete(result);
                    }
                  });
              return future;
            })
        .thenCompose(f -> f);
  }
}
